File: Top interview 150/0053-maximum-subarray/0053-maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        """Kadane's Algorithm은 배열에서 최대 연속 부분합(maximum subarray sum)을 O(n) 시간에 찾는 동적 프로그래밍 알고리즘"""
        max_end = max_curr = nums[0]
        # print(f"init:{max_end, max_curr}")

        for n in nums[1:]:
            # print(f"n = {n}")
            max_end = max(n, n + max_end)
            max_curr = max(max_curr, max_end)
            # print(f"max_end:{max_end}, max_curr:{max_curr} \n")

        return max_curr

File: Top 100 Liked/Misc/0053-maximum-subarray/0053-maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        // Kadane's Algorithm은 배열에서 최대 연속 부분합(maximum subarray sum)을 O(n) 시간에 찾는 동적 프로그래밍 알고리즘
        int maxEnd = nums[0];
        int maxCurr = nums[0];

        for (int i = 1; i < nums.size(); i++)
        {
            int n = nums[i];
            maxEnd = max(n, n + maxEnd);
            maxCurr = max(maxCurr, maxEnd);
        }

        return maxCurr;
    }
};